Add support for ApplicationGateway

vimorra opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

Closes #2990

**What this PR does / why we need it**:
Add support for Application gateway service
**Special notes for your reviewer**:
Still working on the sample
**How does this PR make you feel**:
![gif](https://giphy.com/)

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains tests

matthchr commented on v2/api/network/customizations/network_security_group_extension_test.go:
```diff
@@ -1,70 +0,0 @@
-	"github.com/Azure/azure-service-operator/v2/internal/util/to"
-)
-
-func Test_FuzzySetNetworkSecurityGroupsSecurityRules(t *testing.T) {
```
These files (all of the deletions in this folder) shouldn't have been deleted. Can you revive them?

vimorra replied:
fixed

matthchr commented on v2/api/network/v1api20220701/structure.txt:
```diff
@@ -2,6 +2,1207 @@
+│   │   ├── BackendHttpSettingsCollection: Object (14 properties)[]
+│   │   │   ├── AffinityCookieName: *string
+│   │   │   ├── AuthenticationCertificates: Object (1 property)[]
+│   │   │   │   └── Reference: *genruntime.ResourceReference
```
It would be nice to determine what sort of resource this is, and if we should also add support for it to our list.

matthchr replied:
Ah, it might actually be created by the AuthenticationCertificates section of this resource itself. That will make referring to it a bit annoying as it won't have its own Kubernetes ID. Not that there's much we can do about that.

vimorra replied:
it is quite common in app gateway that internal app gateway resource are linked IP, in this case the BackendHttpSettingsCollection can have a reference to an AuthenticationCertificates, thta is itself a sub resource in application gateway

matthchr commented on v2/api/network/v1api20220701/structure.txt:
```diff
@@ -2,6 +2,1207 @@
+│   │   │   ├── PickHostNameFromBackendAddress: *bool
+│   │   │   ├── Port: *int
+│   │   │   ├── Probe: *Object (1 property)
+│   │   │   │   └── Reference: *genruntime.ResourceReference
```
same, what sort of resource is this? Do we already support it?

matthchr replied:
Ah... it's defined on this resource itself. Weird.

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+		},
+	}
+	publicIPAddress := newPublicIp(tc, testcommon.AsOwner(rg))
+	tc.CreateResourceAndWait(publicIPAddress)
```
minor, you can combined all of these into a single create+wait

`tc.CreateResourcesAndWait(publicIPAddress, vnet, subnet)`

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	appGtwBackendPools, appGtwBackendPoolsID := defineApplicationGatewayBackendAddressPool(tc, rg, appGatewayName)
+	appGtwBackendHttpSettings, appGtwBackendHttpSettingsID := defineApplicationGatewayBackendHttpSettings(tc, rg, appGatewayName)
+	appGtwWafConfig, appGtwWafConfigID := defineApplicationGatewayWebApplicationFirewallConfiguration(tc, rg, appGatewayName)
+	log.Println(appGtwSslProfilesID)
```
use `tc.T.Log` instead so that the logs are correctly interspersed in the parallel test output.

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	tc.Expect(vnet.Status.Id).ToNot(BeNil())
+	tc.CreateResourceAndWait(subnet)
+	tc.Expect(subnet.Status.Id).ToNot(BeNil())
+	appGtsListnerName := "app-gw-http-listner-1"
```
```suggestion
	appGtsListenerName := "app-gw-http-listener-1"
```

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	tc.CreateResourceAndWait(subnet)
+	tc.Expect(subnet.Status.Id).ToNot(BeNil())
+	appGtsListnerName := "app-gw-http-listner-1"
+	httpListnerParams := map[string]string{
```
```suggestion
	httpListenerParams := map[string]string{
```

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+		"param-key2": "httpListeners",
+		"param-val2": appGtsListnerName,
+	}
+	appGtwFHttpListnerID, err := getFrontendPortsARMID(tc, rg, httpListnerParams)
```
minor: You could have this function take 4 args instead of a map of string-string w/ what must be 4 values.

Might be easier to use it then?

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+		RuleSetType:    to.Ptr[string]("OWASP"),
+		RuleSetVersion: to.Ptr[string]("3.2"),
+	})
+	armid_params := map[string]string{
```
```suggestion
	armIDParams := map[string]string{
```

matthchr replied:
same below for all the usages like this. Though I think it'd be cleaner to be a method with 4 args (see my above comment) instead.

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+			SslCertificates: appGtwSslCerts,
+			SslProfiles:     appGtwSslProfiles,
+			SslPolicy: to.Ptr[network.ApplicationGatewaySslPolicy](network.ApplicationGatewaySslPolicy{
+				PolicyName: to.Ptr[network.PolicyNameEnum](network.PolicyNameEnum(network.PolicyNameEnum_AppGwSslPolicy20220101S)),
```
same, don't need `[network.PolicyNameEnum]`, it should be inferred you can just delete the bit in the square brackets

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+			WebApplicationFirewallConfiguration: appGtwWafConfig,
+			HttpListeners: []network.ApplicationGatewayHttpListener{
+				{
+					Name: to.Ptr[string](appGtsListnerName),
```
minor: You don't need to include `[string]` here. It's inferred from the type of the `appGtsListnerName`

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+						},
+					},
+					RuleType: to.Ptr[network.ApplicationGatewayRequestRoutingRulePropertiesFormat_RuleType](network.ApplicationGatewayRequestRoutingRulePropertiesFormat_RuleType(network.ApplicationGatewayRequestRoutingRulePropertiesFormat_RuleType_Basic)),
+					Priority: to.Ptr[int](100),
```
minor: can drop the `[int]` here I believe.

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	subRes := []network.ApplicationGatewaySslCertificate{
+		{
+			//KeyVaultSecretId: to.Ptr[string]("https://keyvaultname.vault.azure.net/secrets/secretname"),
+			Name: to.Ptr[string](subResname),
```
minor: drop `[string]` here too (applies to the whole file)

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+		{
+			//KeyVaultSecretId: to.Ptr[string]("https://keyvaultname.vault.azure.net/secrets/secretname"),
+			Name: to.Ptr[string](subResname),
+			Data: to.Ptr[string](`MIIKEQIBAzCCCdcGCSqGSIb3DQEHAaCCCcgEggnEMIIJwDCCBHcGCSqGSIb3DQEH
```
Is this actually going to pass in our CI subscription? Do we need some magical KeyVault with a secret uploaded to it?

Ideally the test would be self-contained, so it would create (and then tear-down) any resources it needed, including the KV and the secret payload. If it's not doing that we need to think about it.

theunrepentantgeek replied:
If this is an embedded certificate, what's its expiry? 

The test should either be self-contained (creating its own certificate, garanteeing that it's not stale), or at the very least there should be a comment here indicating when the certificate will expire so that a future dev knows what needs doing when the test fails.

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+		{
+			Name: to.Ptr[string](subResname),
+			SslPolicy: &network.ApplicationGatewaySslPolicy{
+				//MinProtocolVersion: to.Ptr[network.ProtocolsEnum](network.ProtocolsEnum(network.ProtocolsEnum_STATUS_TLSv1_2)),
```
minor: commented out code, remove it?

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	subResname := "app-gw-be-pool-1"
+	subRes := []network.ApplicationGatewayBackendHttpSettings{
+		{
+			Name:                           to.Ptr[string](subResname),
```
minor: drop `to.Ptr[...]` (the stuff in square brackets) so it's just `to.Ptr(thing)`.

matthchr commented on v2/samples/network/v1api20220701/v1api20220701_applicationgateway.yaml:
```diff
@@ -0,0 +1,148 @@
+  sslCertificates:
+  - name: app-gw-sslcert-1
+    data: |
+      MIIKEQIBAzCCCdcGCSqGSIb3DQEHAaCCCcgEggnEMIIJwDCCBHcGCSqGSIb3DQEH
```
IMO we should redact this secret, it won't work for users? Or will it? Is this just the publicData or private data too?

vimorra replied:
this is the tls certificate attached to the app gateway to enable tls on the listners, It contains the full chain tls certificates in base64

matthchr commented on v2/samples/network/v1api20220701/v1api20220701_applicationgateway.yaml:
```diff
@@ -0,0 +1,148 @@
+  - name: app-gw-http-listner-1
+    frontendIPConfiguration:
+      reference:  
+        armId: /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/<rgName>/providers/Microsoft.Network/applicationGateways/<appgatewayname>/frontendIPConfigurations/<feIpConfigName>
```
You can replace `<rgname>` with `aso-sample-rg` sincethat's the sample rg which the sample is written in the context of.
Same with `appgatewayname`, it can be `aso-sample-application-gateway`, the name of this application gateway. Use the actual `feIpCOnfigName` too.

Same for all the refs below as well.

vimorra replied:
fixed

matthchr commented on v2/samples/network/v1api20220701/v1api20220701_applicationgateway.yaml:
```diff
@@ -0,0 +1,148 @@
+        reference:  
+          armId: /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/<rgName>/providers/Microsoft.Network/applicationGateways/<appgatewayname>/httpListeners/<httpListnerName>
+      ruleType: Basic
+      priority: 100
```
minor: add newline at end of file.

matthchr commented on v2/samples/network/v1api20220701/v1api20220701_applicationgateway.yaml:
```diff
@@ -0,0 +1,148 @@
+        armId: /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/<rgName>/providers/Microsoft.Network/applicationGateways/<appgatewayname>/frontendIPConfigurations/<feIpConfigName>
+    frontendPort:
+      reference:  
+        armId: /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/<rgName>/providers/Microsoft.Network/applicationGateways/<appgatewayname>/frontendPorts/<fePortName>
```
@super-harsh, is this going to pass sample validation? Do we deal with swapping subid into raw ARM refs?

matthchr commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,389 @@
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD(t *testing.T) {
```
This test needs to be recorded and have the resulting yaml file checked in.

vimorra replied:
here I need your help to understand what is going wrong or in the way a run the test,  I'm running this command:
TIMEOUT=60m AZURE_TENANT_ID=16b3c013-d300-468d-ac64-7eda0820b6d3 AZURE_SUBSCRIPTION_ID=SUB_ID AZURE_CLIENT_ID=ClientID AZURE_CLIENT_SECRET=secrets TEST_FILTER=Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD task controller:test-integration-envtest-record

In the subscription and from the logs everything is working, but the command end with this

[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Reconcile invoked" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" kind="&TypeMeta{Kind:ApplicationGateway,APIVersion:network.azure.com/v1api20220701storage,}" resourceVersion="1153" generation=2 uid="00e2641a-e857-4996-b9a1-4af068dbe6bb" ownerReferences=[]v1.OwnerReference{v1.OwnerReference{APIVersion:"resources.azure.com/v1api20200601storage", Kind:"ResourceGroup", Name:"asotest-rg-xofglu", UID:"e9fda82c-b286-49e1-89e3-d9ab6c010ba6", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}} creationTimestamp="2023-08-23 09:38:47 +0000 UTC" deletionTimestamp="2023-08-23 09:55:22 +0000 UTC" finalizers=[]string{"serviceoperator.azure.com/finalizer"} annotations=map[string]string{"serviceoperator.azure.com/resource-id":"/subscriptions/515ffc0b-a14f-4750-aef5-3db99394b8db/resourceGroups/asotest-rg-xofglu/providers/Microsoft.Network/applicationGateways/appgatewaytest"} conditions="[Condition [Ready], Status = \"True\", ObservedGeneration = 1, Severity = \"\", Reason = \"Succeeded\", Message = \"\", LastTransitionTime = \"2023-08-23 09:43:53 +0000 UTC\"]" owner="asotest-rg-xofglu, Group/Kind: resources.azure.com/ResourceGroup"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Determined Delete action" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest" action="BeginDelete"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Starting delete of resource" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="asotest-publicip-uvwecz"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Delete succeeded, removing finalizer" name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record] 2023/08/23 09:55:23 http: TLS handshake error from 127.0.0.1:56704: EOF
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-publicip-uvwecz resource status: Operation cannot be fulfilled on publicipaddresses.network.azure.com "asotest-publicip-uvwecz": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/publicipaddresses/aso-test-networking-applicationgateway-https-listener-crud/asotest-publicip-uvwecz, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: e483dfb7-ef2b-4557-b595-a422bca35790, UID in object meta: " name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Delete succeeded, removing finalizer" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="asotest-vn-fsazuy"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Delete succeeded, removing finalizer" name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-vn-fsazuy resource status: Operation cannot be fulfilled on virtualnetworks.network.azure.com "asotest-vn-fsazuy": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/virtualnetworks/aso-test-networking-applicationgateway-https-listener-crud/asotest-vn-fsazuy, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 0d3de2d4-a2d4-427b-94b3-0a8b96a7f02c, UID in object meta: " name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/appgatewaytest resource status: Operation cannot be fulfilled on applicationgateways.network.azure.com "appgatewaytest": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/applicationgateways/aso-test-networking-applicationgateway-https-listener-crud/appgatewaytest, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 00e2641a-e857-4996-b9a1-4af068dbe6bb, UID in object meta: " name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record] 2023/08/23 09:55:27 http: TLS handshake error from 127.0.0.1:40460: read tcp 127.0.0.1:34851->127.0.0.1:40460: read: connection reset by peer
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Reconcile invoked" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" kind="&TypeMeta{Kind:VirtualNetworksSubnet,APIVersion:network.azure.com/v1api20201101storage,}" resourceVersion="1159" generation=2 uid="de424115-8da3-4fe6-bb8b-8ec8452e492e" ownerReferences=[]v1.OwnerReference{v1.OwnerReference{APIVersion:"network.azure.com/v1api20201101storage", Kind:"VirtualNetwork", Name:"asotest-vn-fsazuy", UID:"0d3de2d4-a2d4-427b-94b3-0a8b96a7f02c", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}} creationTimestamp="2023-08-23 09:38:32 +0000 UTC" deletionTimestamp="2023-08-23 09:55:27 +0000 UTC" finalizers=[]string{"serviceoperator.azure.com/finalizer"} annotations=map[string]string{"serviceoperator.azure.com/resource-id":"/subscriptions/515ffc0b-a14f-4750-aef5-3db99394b8db/resourceGroups/asotest-rg-xofglu/providers/Microsoft.Network/virtualNetworks/asotest-vn-fsazuy/subnets/AzureAppGtwSubnet"} conditions="[Condition [Ready], Status = \"True\", ObservedGeneration = 1, Severity = \"\", Reason = \"Succeeded\", Message = \"\", LastTransitionTime = \"2023-08-23 09:38:42 +0000 UTC\"]" owner="asotest-vn-fsazuy, Group/Kind: network.azure.com/VirtualNetwork"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Determined Delete action" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet" action="BeginDelete"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Starting delete of resource" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Delete succeeded, removing finalizer" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-subnet-jsrwzs resource status: Operation cannot be fulfilled on virtualnetworkssubnets.network.azure.com "asotest-subnet-jsrwzs": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/virtualnetworkssubnets/aso-test-networking-applicationgateway-https-listener-crud/asotest-subnet-jsrwzs, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: de424115-8da3-4fe6-bb8b-8ec8452e492e, UID in object meta: " name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     kube_per_test_context.go:200: All resources deleted
[controller:test-integration-envtest-record] --- FAIL: Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD (1050.23s)
[controller:test-integration-envtest-record] FAIL
[controller:test-integration-envtest-record] FAIL       github.com/Azure/azure-service-operator/v2/internal/controllers 1051.375s
[controller:test-integration-envtest-record] FAIL
task: Failed to run task "controller:test-integration-envtest-record": exit status 1

matthchr replied:
We can take a  look and help with this bit.

theunrepentantgeek replied:
Reading through the log, I found something interesting.

The version first quoted log line seems to indicate the ApplicationGateway was successfully created:

```
I2023-08-23T09:55:22Z] ApplicationGatewayController 
Condition [Ready], Status = "True", ObservedGeneration = 1, Severity = "", Reason = "Succeeded", Message = "", LastTransitionTime = "2023-08-23 09:43:53 +0000 UTC"
```
Next, we see the test start to clean up:
```
I2023-08-23T09:55:22Z] ApplicationGatewayController
Starting delete of resource
```
But, the last log line is just 6 seconds later, which is way too fast for it to have successfully cleaned up by deleting everything:
```
I2023-08-23T09:55:28Z
```
I'll run the test locally to repro.

vimorra commented on v2/api/network/v1api20220701/structure.txt:
```diff
@@ -2,6 +2,1207 @@
+│   ├── Owner: github.com/Azure/azure-service-operator/v2/api/resources/v1apiv20191001.ResourceGroup
+│   ├── Spec: Object (37 properties)
+│   │   ├── AuthenticationCertificates: Object (2 properties)[]
+│   │   │   ├── Data: *string
```
This is the public key certificate that is uploaded to enable mtls on the application gateway, this is used by the app gateway to trust the client

theunrepentantgeek commented on v2/azure-arm.yaml:
```diff
@@ -1561,6 +1561,51 @@ objectModelConfiguration:
     2022-07-01:
+      ApplicationGateway:
+        $export: true
+        $supportedFrom: v2.3.0
```
I don't think this will make it for v2.3.0.

```suggestion
        $supportedFrom: v2.4.0
```

vimorra replied:
fixed in updated commit

theunrepentantgeek commented on v2/azure-arm.yaml:
```diff
@@ -1561,6 +1561,51 @@ objectModelConfiguration:
+      ApplicationGatewayBackendHttpSettings:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayBackendSettings:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayFrontendIPConfiguration:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayFrontendPort:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayIPConfiguration:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayHttpListener:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayListener:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayLoadDistributionPolicy:
+         $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayPrivateLinkConfiguration:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayProbe:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayRedirectConfiguration:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayRequestRoutingRule:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayRewriteRuleSet:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayRoutingRule:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewaySslCertificate:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewaySslProfile:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayTrustedClientCertificate:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayTrustedRootCertificate:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayUrlPathMap:
+        $resourceEmbeddedInParent: ApplicationGateway
+      ApplicationGatewayAuthenticationCertificate:
+        $resourceEmbeddedInParent: ApplicationGateway
```
Can any of these be created independently, and then referenced by the parent? 

If so, consider using `$isResource` instead to make these available as independent resources. Amongst other factors, this would make the users yaml file easier to read.

vimorra replied:
As shared, the embedded resources are tightly coupled with Appgateway, so my suggestion is to maintain the current implementation.

theunrepentantgeek reviewed (commented):
Still a few cleanup items required (see comments by @matthchr) - including more reliance on Go type inference, which will clean up the test in `controllers` some. I'll have a look at the test and work out what the failure means.

theunrepentantgeek reviewed (changes requested):
We're keen to get this PR merged, it's looking very close to finished.

There are just two items outstanding:
* The signature of `getFrontendPortsARMID()` needs to be modified to avoid the brittle nature of passing around arbitrary maps.
* We can't commit secrets to the repo, so we need to tidy up the use of certificates.

I believe we have infrastructure available through the TestContext (`tc`) to create the required secrets on the fly, so that nothing needs to be hard coded. If you find the capability you need is missing, let us know and we'll fill in the gap for you.

matthchr commented:
If you're busy @vimorra and don't have the time to finish updating based on our feedback, we can pick this up and get it over the line. No promises that it makes it into 2.4.0 but we'll try.

vimorra commented on v2/internal/controllers/crd_networking_applicationgateway_test.go:
```diff
@@ -0,0 +1,394 @@
+	return subRes, subResARMID
+}
+
+func getFrontendPortsARMID(tc *testcommon.KubePerTestContext, rg *resources.ResourceGroup, params map[string]string) (string, error) {
```
fixed with option 1. @theunrepentantgeek  please review

vimorra commented:
> If you're busy @vimorra and don't have the time to finish updating based on our feedback, we can pick this up and get it over the line. No promises that it makes it into 2.4.0 but we'll try.

yes I will do my best, fixed the comment 1 about the signature of getFrontendPortsARMID. I will try to fix the certificate issue by adding the isSecret

vimorra commented:
> We're keen to get this PR merged, it's looking very close to finished.
> 
> There are just two items outstanding:
> 
> * The signature of `getFrontendPortsARMID()` needs to be modified to avoid the brittle nature of passing around arbitrary maps.
> * We can't commit secrets to the repo, so we need to tidy up the use of certificates.
> 
> I believe we have infrastructure available through the TestContext (`tc`) to create the required secrets on the fly, so that nothing needs to be hard coded. If you find the capability you need is missing, let us know and we'll fill in the gap for you.

not clear which option you are suggesting if create the certificate in an automated way or mark the attribute as "isSecret" ? in any case con you provide an example @theunrepentantgeek

theunrepentantgeek commented:
I went back and had another look, and it's actually both of the issues you've identified.

The issue I was referring to above is about hard coded certificates:
* the file `appgwcert64.pfx`
* the file `appgwcert.pfx`
* the `sslCertificates.data` block in the `sample v1api20220701_applicationgateway.yaml`

Checking certificates into repos is a security issue _even if_ they only contain public keys. They also expire, which will result in the test failing in the future. 

The second issue is the `authenticationCertificates.Data` property on `ApplicationGateway.spec` - this should be configured with `$isSecret: true` so that the value is pulled from the cluster.  Apologies for missing this earlier.

This will have two benefits. 
* When the certificate is rotated, updating the secret will result in ASO updating the ApplicationGateway, much easier (and more automated) than having to redeploy everything from scratch. (Rotating the certificate directly in Azure isn't an option because ASO will proactively restore the configured certificate next time it reconciles the resource.) 
* The certificate won't be disclosed to someone looking at the YAML for the resource. If the cert contains only a public key, this wouldn't matter - but people do take shortcuts and make mistakes, and if they use a cert containing a private key disclosure is a security risk. 

We believe there's infrastructure available on the `TestContext` used in your test that should allow you to create a certificate on the fly as a part of the test. If this is missing, or if you can't find it, ping us and we'll dig in further.

vimorra commented:
@matthchr, @super-harsh and @theunrepentantgeek I need your support to understand how to generate an on the fly ssl certificate.
Meanwhile I have fixed the:

- $isSecret: true point on all the resources that point to secrets and certificates
- fixed the signature of $isSecret: true

theunrepentantgeek commented:
@vimorra - I'm looking into generating the SSL certificate; hoping to put together a sample for ASO users that will be relevant to your needs too. I'll update back here as that progresses.

vimorra commented:
Hi @theunrepentantgeek  and @matthchr  with the support of @super-harsh the PR should be ready to be merged 
, shall we look together at how to solve the conflicts or I can assume that the incoming version is the one to merge?

super-harsh commented:
@vimorra An easy way to deal with these conflicts is to merge `main` into your branch and run `task controller:generate-types` task and commit the files.

theunrepentantgeek commented:
/ok-to-test sha=d1ebad0

theunrepentantgeek commented:
/azp run
